File: src/apps/daemon/internal/networking/dns/dns_fragment_store.go
// Package dns provides networking, tunneling, and scheduling primitives.
// Conforms strictly to §8 structural cleanroom rules.

package dns

import (
	"sync"
	"time"
)

type fragmentEntry struct {
	createdAt      time.Time
	totalFragments uint8
	chunks         [256][]byte
	count          uint8
}

// DnsFragmentStore collects multi-chunk DNS tunnel datagrams and reassembles them upon completion.
// Includes duplicate completion tracking with configurable TTL retention.
type DnsFragmentStore[K comparable] struct {
	mu        sync.Mutex
	items     map[K]*fragmentEntry
	completed map[K]time.Time
	lastPurge time.Time
}

// NewDnsFragmentStore creates an initialized store with given initial capacity.
func NewDnsFragmentStore[K comparable](capacity int) *DnsFragmentStore[K] {
	if capacity < 1 {
		capacity = 16
	}
	return &DnsFragmentStore[K]{
		items:     make(map[K]*fragmentEntry, capacity),
		completed: make(map[K]time.Time, capacity),
		lastPurge: time.Now(),
	}
}

// Collect processes a received fragment for a message key.
// Returns:
// - assembled payload (non-nil when all chunks are collected)
// - completed boolean (true when assembled or single-packet succeeded)
// - duplicate boolean (true if packet was already delivered within retention period)
func (s *DnsFragmentStore[K]) Collect(key K, payload []byte, fragmentID uint8, totalFragments uint8, now time.Time, retention time.Duration) ([]byte, bool, bool) {
	// Single fragment fast path
	if totalFragments <= 1 {
		if retention <= 0 {
			return append([]byte(nil), payload...), true, false
		}

		s.mu.Lock()
		defer s.mu.Unlock()

		s.maybePurgeLocked(now, retention)
		if expiresAt, ok := s.completed[key]; ok && now.Before(expiresAt) {
			return nil, false, true
		}

		delete(s.items, key)
		s.completed[key] = now.Add(retention)
		return append([]byte(nil), payload...), true, false
	}

	if fragmentID >= totalFragments {
		return nil, false, false
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	s.maybePurgeLocked(now, retention)
	if expiresAt, ok := s.completed[key]; ok && now.Before(expiresAt) {
		return nil, false, true
	}

	current, ok := s.items[key]
	if !ok || current.totalFragments != totalFragments {
		current = &fragmentEntry{
			createdAt:      now,
			totalFragments: totalFragments,
		}
		s.items[key] = current
	}

	if current.chunks[fragmentID] == nil {
		current.count++
	}
	current.chunks[fragmentID] = append(current.chunks[fragmentID][:0], payload...)

	if current.count < totalFragments {
		return nil, false, false
	}

	// All fragments received: assemble contiguous buffer
	totalSize := 0
	for i := uint8(0); i < totalFragments; i++ {
		chunk := current.chunks[i]
		if chunk == nil {
			return nil, false, false
		}
		totalSize += len(chunk)
	}

	assembled := make([]byte, 0, totalSize)
	for i := uint8(0); i < totalFragments; i++ {
		assembled = append(assembled, current.chunks[i]...)
	}

	delete(s.items, key)
	if retention > 0 {
		s.completed[key] = now.Add(retention)
	} else {
		delete(s.completed, key)
	}

	return assembled, true, false
}

// Purge evicts expired entries and completed sequence markers.
func (s *DnsFragmentStore[K]) Purge(now time.Time, retention time.Duration) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.purgeLocked(now, retention)
}

func (s *DnsFragmentStore[K]) maybePurgeLocked(now time.Time, retention time.Duration) {
	if now.Sub(s.lastPurge) >= time.Second {
		s.purgeLocked(now, retention)
		s.lastPurge = now
	}
}

func (s *DnsFragmentStore[K]) purgeLocked(now time.Time, retention time.Duration) {
	maxAge := retention
	if maxAge < 5*time.Second {
		maxAge = 5 * time.Second
	}

	for k, entry := range s.items {
		if now.Sub(entry.createdAt) > maxAge {
			delete(s.items, k)
		}
	}
	for k, exp := range s.completed {
		if now.After(exp) {
			delete(s.completed, k)
		}
	}
}

// ActiveEntries returns the count of currently uncompleted fragment sets.
func (s *DnsFragmentStore[K]) ActiveEntries() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.items)
}

File: src/packages/lumicore/src/transport/dns_fragment_store.rs
//! # DNS Packet Fragmentation Store
//!
//! Collects and reassembles fragmented UDP/DNS tunnel datagrams into complete contiguous packets.
//! Provides single-fragment fast-path, retention-based duplicate suppression, and periodic expiration.
//! Conforms to §8 structural cleanroom rules.

use std::collections::HashMap;
use std::hash::Hash;
use std::sync::Mutex;
use std::time::{Duration, Instant};

struct FragmentEntry {
    created_at: Instant,
    total_fragments: u8,
    chunks: Vec<Option<Vec<u8>>>,
    received_count: u8,
}

/// Thread-safe fragment collector and reassembly store.
pub struct DnsFragmentStore<K: Eq + Hash + Clone> {
    state: Mutex<StoreState<K>>,
}

struct StoreState<K> {
    items: HashMap<K, FragmentEntry>,
    completed: HashMap<K, Instant>,
    last_purge: Instant,
}

#[derive(Debug, PartialEq, Eq)]
pub enum CollectResult {
    /// Packet is partially received; waiting for more fragments.
    Incomplete,
    /// Packet was previously completed and is rejected within the retention window.
    DuplicateSuppressed,
    /// All fragments have been collected and contiguous payload reassembled.
    Assembled(Vec<u8>),
}

impl<K: Eq + Hash + Clone> DnsFragmentStore<K> {
    pub fn new(capacity: usize) -> Self {
        Self {
            state: Mutex::new(StoreState {
                items: HashMap::with_capacity(capacity),
                completed: HashMap::with_capacity(capacity),
                last_purge: Instant::now(),
            }),
        }
    }

    /// Collects a fragment for the given message key.
    /// - If `total_fragments <= 1`, fast path returns data immediately while remembering completion.
    /// - When all fragments `0..total_fragments-1` are collected, returns `CollectResult::Assembled`.
    pub fn collect(
        &self,
        key: K,
        payload: &[u8],
        fragment_id: u8,
        total_fragments: u8,
        retention: Duration,
    ) -> CollectResult {
        let now = Instant::now();

        // Single fragment fast path
        if total_fragments <= 1 {
            if retention.is_zero() {
                return CollectResult::Assembled(payload.to_vec());
            }

            let mut state = self.state.lock().unwrap();
            Self::maybe_purge(&mut state, now, retention);

            if let Some(&expires_at) = state.completed.get(&key) {
                if now < expires_at {
                    return CollectResult::DuplicateSuppressed;
                }
            }

            state.items.remove(&key);
            state.completed.insert(key, now + retention);
            return CollectResult::Assembled(payload.to_vec());
        }

        if fragment_id >= total_fragments {
            return CollectResult::Incomplete;
        }

        let mut state = self.state.lock().unwrap();
        Self::maybe_purge(&mut state, now, retention);

        if let Some(&expires_at) = state.completed.get(&key) {
            if now < expires_at {
                return CollectResult::DuplicateSuppressed;
            }
        }

        let entry = state.items.entry(key.clone()).or_insert_with(|| FragmentEntry {
            created_at: now,
            total_fragments,
            chunks: vec![None; total_fragments as usize],
            received_count: 0,
        });

        if entry.total_fragments != total_fragments {
            // Sequence mismatch: reset entry
            entry.total_fragments = total_fragments;
            entry.chunks = vec![None; total_fragments as usize];
            entry.received_count = 0;
            entry.created_at = now;
        }

        let idx = fragment_id as usize;
        if idx < entry.chunks.len() {
            if entry.chunks[idx].is_none() {
                entry.received_count += 1;
            }
            entry.chunks[idx] = Some(payload.to_vec());
        }

        if entry.received_count < total_fragments {
            return CollectResult::Incomplete;
        }

        // All fragments have arrived: assemble contiguous buffer
        let mut total_len = 0;
        for chunk_opt in &entry.chunks {
            match chunk_opt {
                Some(chunk) => total_len += chunk.len(),
                None => return CollectResult::Incomplete,
            }
        }

        let mut assembled = Vec::with_capacity(total_len);
        for chunk_opt in &entry.chunks {
            if let Some(chunk) = chunk_opt {
                assembled.extend_from_slice(chunk);
            }
        }

        state.items.remove(&key);
        if !retention.is_zero() {
            state.completed.insert(key, now + retention);
        }

        CollectResult::Assembled(assembled)
    }

    /// Purges expired incomplete entries and completion markers.
    pub fn purge(&self, retention: Duration) {
        let now = Instant::now();
        let mut state = self.state.lock().unwrap();
        Self::purge_internal(&mut state, now, retention);
    }

    fn maybe_purge(state: &mut StoreState<K>, now: Instant, retention: Duration) {
        if now.duration_since(state.last_purge) >= Duration::from_secs(1) {
            Self::purge_internal(state, now, retention);
            state.last_purge = now;
        }
    }

    fn purge_internal(state: &mut StoreState<K>, now: Instant, retention: Duration) {
        let max_age = if retention > Duration::from_secs(5) {
            retention
        } else {
            Duration::from_secs(5)
        };

        state.items.retain(|_, entry| now.duration_since(entry.created_at) < max_age);
        state.completed.retain(|_, &mut expires_at| now < expires_at);
    }

    /// Returns the number of currently active incomplete fragment sets.
    pub fn active_count(&self) -> usize {
        let state = self.state.lock().unwrap();
        state.items.len()
    }
}
